docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,546 @@
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
+
+* Validation will be present that only one of existing spec field OR new reference
+field is present in custom resource, not both.
+
+* If the existing spec field was marked as required, validation will be added that
+at least one of existing spec field or new reference field is present.
+
+* If `AWSResourceReference` type field is present in the desired state, first ACK
+controller will query for the referenced resource. If the referenced resource
+has `ACK.ResourceSynced` condition status `True` & specified field is found
+in the referenced resource, the reconciliation loop will progress. Otherwise
+the request will get requeued for future processing.
+
+* Watchers for the referenced resource type will not be implemented in the initial
+implementation. The initial implementation will mostly be used for finding
+identifier of referenced resource and any update in referenced resource should not
+change it's identifier. As mentioned earlier, deletion will also be handled in
+future.
+
+* The code to determine when the referenced resource is synced will be generated,
+and additional hooks will be added for customization. Ex: Do not just check presence
+of a field but also check if the Status is available etc...
+
+### Low Level Implementation
+
+#### generator.yaml Changes
+
+```go
+type FieldConfig struct {
+    ...
+    Reference *ReferenceConfig `json:"references,omitempty"`
+    ...
+}
+
+// ReferenceConfig contains the instructions for how to add the referenced resource
+// configuration for a field.
+type ReferenceConfig struct {
+	// Kind mentions the K8s Kind of referenced resource
+	Kind string `json:"kind"`
+	// FieldPath refers to the the path of field which should be copied
+	// from referenced resource into TargetFieldName
+	FieldPath string `json:"field_path"`
+	// TargetFieldName is spec field name which gets value from
+	// referenced resource's FieldPath
+	TargetFieldName string `json:"target_field"`
+	// IsList mentions whether the new field should be AWSResourceReference or
+	// List of AWSResourceReference
+	IsList          bool   `json:"is_list"`
+	// Required mentions whether either of AWSResourceReference or
+	// TargetFieldName must be present in custom resource
+	Required        bool   `json:"required"`
+}
+```
+
+`APIGatewayv2 generator.yaml` Example:
+```yaml
+resources:
+  Integration:
+    fields:
+      Api:
+        references:
+          kind: API
+          field_path: Status.APIID
+          target_field: APIID
+          is_list: false
+          required: true
+      ApiId:
+        is_required: false
+```
+NOTE: `ApiId` field is marked as not required because the new reference field
+`Api` can also provide APIId.
+
+#### New AWSResourceReference Type
+
+```go
+type AWSResourceReference struct {
+	Name *string `json:"name,omitempty"`
```

Comment:
We could also name this field `From` so that YAML would look like:

```yaml
spec:
  apiIDRef:
    from: my-api-name
```

Reply:
Looking at our discussion notes from the other week, I thought we decided to go with:
```yaml
spec:
  apiIDRef:
    from:
      name: my-api-name
```

Leaving the door open for a `from.labelSelector` or otherwise?

Reply:
@jaypipes , will using `from` instead of `name` create confusion when in future we add labelSelector. The example that @RedbackThomson provided seems like better use of 'from' keyword while generating the manifest. wdyt?

--------------

To implement `from` keyword, how does the following look? (Note the new struct we will need to add here.)

```go
type IntegrationSpec struct {
	APIID    *string                           `json:"apiID,omitempty"`
	APIIDRef *ackv1alpha1.ReferenceProvider `json:"apiIDRef,omitempty"`
...
}

type ReferenceProvider struct {
        ResourceReference *AWSResourceReference `json:"from,omitempty"`
}

type AWSResourceReference struct {
	Name *string `json:"name,omitempty"`
}
```

Reply:
@vijtrip2 yeah, agreed. @RedbackThomson's example shows a better use of From.